src/sqlfluff/utils/reflow/reindent.py

```diff
@@ -1546,6 +1537,126 @@ def _match_indents(
+        e = cast(ReflowPoint, elements[e_idx])
+        # We need to check for negative sections so they get the right
+        # indent (otherwise they'll be over indented).
+        # The `desired_indent` above is for the "uphill" side.
+        following_class_types = elements[e_idx + 1].class_types
+        indent_stats = e.get_indent_impulse(
+            allow_implicit_indents, following_class_types
+        )
+        # Cache them for later
+        _indent_stats_cache[e_idx] = indent_stats
+
+    # If we can get to the uphill indent of later break, and still be within
+    # the line limit, then we can skip everything before it.
+    purge_before = 0
+    for e_idx in target_breaks:
+        # Is the following block already past the limit?
+        # NOTE: We use the block because we know it will have segments.
+        if not elements[e_idx + 1].segments[0].pos_marker:
+            # If it doesn't have position - we should just bow out
+            # now. It's too complicated.
+            break  # pragma: no cover
+        if (
+            elements[e_idx + 1].segments[0].pos_marker.working_line_pos
+            > line_length_limit
+        ):
+            # If we're past the line length limit, stop looking.
+            break
+
+        # Fetch cached indent stats
+        indent_stats = _indent_stats_cache[e_idx]
+        if indent_stats.trough < 0:
+            # It's negative. Skip onward.
+            continue
+
+        # If we get this far, then it's positive, but still within
+        # the line limit. We can purge any pairs before this.
+        purge_before = e_idx
+        reflow_logger.debug("    ...breaks before %s unnecessary.", purge_before)
+    # Only keep indices which are after the critical point.
+    target_breaks = [e_idx for e_idx in target_breaks if e_idx >= purge_before]
+    reflow_logger.debug("    Remaining breaks: %s.", target_breaks)
+
+    for e_idx in target_breaks:
+        e = cast(ReflowPoint, elements[e_idx])
+        indent_stats = _indent_stats_cache[e_idx]
+        if indent_stats.trough < 0:
+            new_indent = outer_indent
+            # NOTE: If we're about to insert a dedent before a
+            # comma or semicolon ... don't. They are a bit special
+            # in being allowed to trail.
+            if elements[e_idx + 1].class_types.intersection(
+                ("statement_terminator", "comma")
+            ):
+                reflow_logger.debug("    Skipping dedent before comma or semicolon.")
+                # We break rather than continue because this is
+                # necessarily a step back down.
+                break
+        else:
+            new_indent = inner_indent
+
+        new_results, new_point = e.indent_to(
+            new_indent,
+            after=elements[e_idx - 1].segments[-1],
+            before=elements[e_idx + 1].segments[0],
+        )
+        # NOTE: Mutation of elements.
+        elements[e_idx] = new_point
```

Comment:
Same question about mutation vs returning a list.

Reply:
I've updated it to only mutate and not return 👍